model/riscv_insts_cmoext.sail

```diff
@@ -0,0 +1,245 @@
+  match cbop {
+    CBO_CLEAN => {
+      if (check_clean_flush(cur_privilege)) then
+        /* perform clean*/
+        process_clean_inval(vaddr, cache_block_size, true, false)
+      else {
+        handle_illegal();
+        RETIRE_FAIL
+      }
+    },
+    CBO_FLUSH => {
+      if (check_clean_flush(cur_privilege)) then
+        /* perform flush*/
+        process_clean_inval(vaddr, cache_block_size, true, true)
+      else {
+        handle_illegal();
+        RETIRE_FAIL
+      }
+    },
+    CBO_INVAL => {
+      match (check_inval(cur_privilege)) {
+        0b00 => {
+          handle_illegal();
+          RETIRE_FAIL
+          },
+        0b01 => process_clean_inval(vaddr, cache_block_size, true, true),  /*perform flush*/
+        0b10 => process_clean_inval(vaddr, cache_block_size, false, true) /*perform inval*/
+      }
+    },
+    CBO_ZERO  => {
+      if (check_zero(cur_privilege)) then
+        process_cbo_zero(vaddr, cache_block_size)
+      else {
+        handle_illegal();
+        RETIRE_FAIL
+      }
+    }
+  }
+}
+
+union clause ast = RISCV_PREFETCH : (preop, regidx, bits(7))
+
+mapping encdec_prefetch : preop <-> bits(5) = {
+  PREFETCH_I <-> 0b00000,
+  PREFETCH_R <-> 0b00001,
+  PREFETCH_W <-> 0b00011
+}
+
+mapping clause encdec = RISCV_PREFETCH(op, rs1, offset)
+  <-> offset @ encdec_prefetch(op) @ rs1 @ 0b110 @ 0b00000 @ 0b0010011
+
+mapping pre_mnemonic : preop <-> string = {
+  PREFETCH_I <-> "prefetch.i",
+  PREFETCH_R <-> "prefetch.r",
+  PREFETCH_W <-> "prefetch.w"
+}
+
+mapping clause assembly = RISCV_PREFETCH(op, rs1, offset)
+  <-> pre_mnemonic(op) ^ spc() ^ hex_bits_12(offset @ 0b00000) ^ "(" ^ reg_name(rs1) ^ ")"
+
+function clause execute(RISCV_PREFETCH(preop, rs1, offset)) = {
+  /* do nothing other than translate_address for prefetch.i/r/w */
+  match preop {
+    PREFETCH_I => {
+      let vaddr = X(rs1) + EXTS(offset @ 0b00000);
+      match ext_control_check_addr(vaddr) {
+        Ext_ControlAddr_Error(e)  => RETIRE_SUCCESS,
+          Ext_ControlAddr_OK(vaddr) =>
+            match translateAddr(vaddr, Execute()) {
+              TR_Failure(e, _)    => RETIRE_SUCCESS,
+              TR_Address(paddr, _) => RETIRE_SUCCESS
+            }
+      }
+    },
+    PREFETCH_W => {
+      match ext_data_get_addr(rs1, EXTS(offset @ 0b00000), Write(Data), BYTE) {
+        Ext_DataAddr_Error(e)  => RETIRE_SUCCESS,
+        Ext_DataAddr_OK(vaddr) =>
+          match translateAddr(vaddr, Write(Data)) {
+            TR_Failure(e, _)     => RETIRE_SUCCESS,
+            TR_Address(paddr, _) => RETIRE_SUCCESS
+          }
+      }
+    },
+    PREFETCH_R => {
+      match ext_data_get_addr(rs1, EXTS(offset @ 0b00000), Read(Data), BYTE) {
+        Ext_DataAddr_Error(e)  => RETIRE_SUCCESS,
+        Ext_DataAddr_OK(vaddr) =>
+          match translateAddr(vaddr, Read(Data)) {
+            TR_Failure(e, _)     => RETIRE_SUCCESS,
+            TR_Address(paddr, _) => RETIRE_SUCCESS
+          }
+      }
+    }
+  }
+}
```

Comment:
Make sure every line has a terminating newline character (required by POSIX for a file to be a valid text file)